### DISCUSSION TITLE
How can I get the current value of shared.gradio['<some key>']?

### DISCUSSION THREAD

**Discussion by ClayShoaf** [oobabooga/text-generation-webui#1154] [Q&A] [answered]
I'm trying to make an extension, and if I try to pull information from the `shared.gradio` dictionary, I can only retrieve data from the initial creation of `shared.gradio`. The only way I have found to get the current key value is to create a button in `ui()` and then do something like: `my_button.click(my_function, shared.gradio['name1'], my_output)`. Unfortunately, this is very limited, because if I try to do something like `my_button.click(my_function, shared.gradio[my_dropdown], my_output)`, it will only use the initial value of `my_dropdown` instead of the current value. At this point, I am pretty much at a loss.

**Comment by ClayShoaf** [comment]
https://user-images.githubusercontent.com/31578812/232067381-b0f6fe41-8082-4dce-b034-39fc4d227384.mp4

Is this a problem with gradio? It feels like gradio is doing this to me.

**Comment by dibrale** [comment]
I've been struggling with Gradio myself as I write extensions. The Gradio interface is very limiting. I recall reading in a discussion that this is by design, in order to avoid scope creep. However, it does make doing anything beyond basic functions a challenge. I've been working around it by separately tracking objects and/or parameters outside of Gradio when possible.

**Reply by ClayShoaf** [reply]
Yea, that's what I ended up doing. You can make a button that grabs everything with `.then()` statements and then throw them to functions that set global parameters. Kind of a tedious workaround, and so far I haven't found any way to keep things up to date without some input from the user.

**Comment by bmoconno** [comment, marked as answer]
A little more context might help us help you, but another method you can use is to set up `.change` events on the `shared.gradio[...]` elements that you want in the `ui` method and save those into your own `dict`:
```
def ui():
    global backup
    shared.gradio['name1'].change(lambda x: backup.update({'name1': x}), shared.gradio['name1'], [])
    shared.gradio['name2'].change(lambda x: backup.update({'name2': x}), shared.gradio['name2'], [])
    return None
```

And just define your `dict` somewhere like:
```
backup = {
    'name1': None, 
    'name2': None
}
```

Then when either of those values change, your extension will have the latest value.

Probably functionally very similar to a bunch of `.then` statements, but maybe easier to read and doesn't require a button... and you could probably programmatically generate all the `.change` events and your own `dict` with a list of strings for the variables you care about.

**Reply by ClayShoaf** [reply]
I think this might be just what I'm looking for! But do those entries in `shared.gradio` "change" back to their initialization state after `server.py` or `chat.py` does whatever it's doing with generations?

**Reply by bmoconno** [reply]
The `.change` method is only called when the value in that gradio element is updated, the `chat.py` doesn't directly update anything, it's functions are called from the `server.py` as a result of, for example, clicking the `Generate` button, which calls this gradio code:

```
gen_events.append(shared.gradio['Generate'].click(
    lambda x: (x, ''), shared.gradio['textbox'], [shared.gradio['Chat input'], shared.gradio['textbox']], show_progress=False).then(
    chat.cai_chatbot_wrapper, shared.input_params, shared.gradio['display'], show_progress=shared.args.no_stream).then(
    chat.save_history, shared.gradio['Chat mode'], None, show_progress=False)
)
```
Which does the following things in this order:
* Saves the value in the input `textbox` to the `Chat input` state variable and clears out the `textbox` (this is so the text a user just entered goes away when they hit generate instead of staying there until the reply is finished generating).
* Calls the `chat.cai_chatbot_wrapper` method and saves it's result to the `display` textbox.
* Calls the `save_history` method which writes the current version of the chat log to the persistent `.json` file for the selected character.

So, the only `shared.gradio` value that will change after a chat happens is the `display` textbox which will now be updated to show the latest reply from the bot.

**Reply by ClayShoaf** [reply]
@Brawlence this is what you're looking for to get the elusive `name1`
